subt/radio.py

```diff
@@ -33,52 +35,55 @@ def draw_positions(arr):
 class Radio(Node):
     def __init__(self, config, bus):
         super().__init__(config, bus)
-        bus.register('radio', 'artf_xyz', 'breadcrumb')
-        self.min_transmit_dt = timedelta(seconds=10)  # TODO config?
+        bus.register('radio', 'artf_xyz', 'breadcrumb', 'robot_xyz')
+        self.min_transmit_dt = 0  # i.e. every sim_time_sec -> every simulated second
         self.last_transmit = None
+        self.sim_time_sec = None
         self.recent_packets = []
         self.verbose = config.get('verbose', False)
         self.debug_robot_poses = []
+        self.message_counter = 0
 
-    def send_data(self, data):
-        self.last_transmit = self.publish('radio', data + b'\n')
-        return self.last_transmit
+    def send_data(self, channel, data):
+        raw = serialize([self.message_counter, channel, data])
+        self.publish('radio', raw)
+        self.message_counter += 1
 
     def on_radio(self, data):
         src, packet = data
         name = src.decode('ascii')
-        if packet.startswith(b'['):
-            arr = literal_eval(packet.decode('ascii'))
-            if len(arr) == 3:
-                # position
-                if self.verbose:
-                    self.debug_robot_poses.append((self.time, name, arr))  # TODO sim_time_sec
-            elif len(arr) == 4:
-                # artifact
-                self.publish('artf_xyz', [arr])  # publish also standard "list" of detected artifacts
-            else:
-                assert False, arr  # unexpected size/type
-        if packet.startswith(b'{'):
-            arr = literal_eval(packet.decode('ascii'))
-            for k, v in arr.items():
-                assert k in ['breadcrumb', ], k
-                self.publish(k, v)
+        __, channel, msg_data = deserialize(packet)
+        if channel == 'artf':
+            self.publish('artf_xyz', msg_data)  # topic rename - beware of limits 1500bytes!
+        elif channel == 'breadcrumb':
+            self.publish(channel, msg_data)
+        elif channel == 'xyz':
+            self.publish('robot_xyz', [name, msg_data])
+            if self.verbose:
+                self.debug_robot_poses.append((self.time, name, msg_data))
 
     def on_breadcrumb(self, data):
-        self.send_data(bytes(str({'breadcrumb':data}), encoding='ascii'))
+        self.send_data('breadcrumb', data)
+
+    def on_pose3d(self, data):
+        if self.sim_time_sec is None:
+            return
+        if self.last_transmit is None or self.sim_time_sec > self.last_transmit + self.min_transmit_dt:
+            self.send_data('xyz', [self.sim_time_sec, data[0]])
+            self.last_transmit = self.sim_time_sec
+
+    def on_artf(self, data):
+        self.send_data('artf', data)
+
+    def on_sim_time_sec(self, data):
+        self.sim_time_sec = data  # duplicate for super().update(), used just for easier unittesting
+        self.send_data('sim_time_sec', data)
```

Comment:
Why are we broadcasting sim_time_sec?

Reply:
Well, this was a bug with Teambase, which is (and used to be) broadcasting `sim_time_sec`. I can change it to parameter in `config` if you want?

Reply:
I'd rather agree on the high level stuff first.